File: tm/NativeSampleModule.ts
import type {TurboModule} from 'react-native/Libraries/TurboModule/RCTExport';

import {TurboModuleRegistry} from 'react-native';

export interface Spec extends TurboModule {
  readonly reverseString: (input: string) => string;
  readonly passLargeNumber: (input: string) => number;
}

export default (TurboModuleRegistry.getEnforcing<Spec>('NativeSampleModule'));

File: tm/NativeSampleModule.cpp
#include "NativeSampleModule.h"
#include <cmath>

namespace facebook::react {

NativeSampleModule::NativeSampleModule(std::shared_ptr<CallInvoker> jsInvoker)
    : NativeSampleModuleCxxSpec(std::move(jsInvoker)) {}

std::string NativeSampleModule::reverseString(jsi::Runtime& rt, std::string input) {
  return std::string(input.rbegin(), input.rend());
}

int32_t NativeSampleModule::passLargeNumber(jsi::Runtime& rt, int64_t input) {
    return std::cbrt(input);
}

CustomType NativeSampleModule::passCustomType(jsi::Runtime& rt, CustomType input) {
  input.key = "1909";
  input.enabled = !input.enabled;
  input.time = 42;
  return input;
}

} // namespace facebook::react
